src/dice/UpperDice.rb

```diff
@@ -1,157 +1,167 @@
+    @target_number = @cmp_op ? m[5].to_i : nil
+    @reroll_threshold = reroll_threshold(m[2] || m[6])
 
-    output = "#{totalDiceString}#{formatBonus(bonusValue)}"
+    @modify_number = m[3] ? ArithmeticEvaluator.new.eval(m[3], @diceBot.fractionType.to_sym) : 0
 
-    maxValue = maxDiceValue + bonusValue
-    totalValue += bonusValue
-
-    string += "[#{@upper}]" + modifier
-
-    if @diceBot.isPrintMaxDice && (totalDiceCount > 1)
-      output = "#{output} ＞ #{totalValue}"
+    if @reroll_threshold <= 1
+      return "#{@nick_e}: (#{expr()}) ＞ 無限ロールの条件がまちがっています"
     end
 
-    if @signOfInequality != ""
-      output = "#{output} ＞ 成功数#{totalSuccessCount}"
-      string += "#{@signOfInequality}#{diff}"
-    else
-      output += getMaxAndTotalValueResultStirng(maxValue, totalValue, totalDiceCount)
+    roll_list = []
+    @command.split('+').each do |u|
+      times, sides = u.split("U", 2).map(&:to_i)
+      roll_list.concat(roll(times, sides))
     end
 
-    output = "#{@nick_e}: (#{string}) ＞ #{output}"
-
-    if output.length > $SEND_STR_MAX
-      output = "#{@nick_e}: (#{string}) ＞ ... ＞ #{totalValue}"
-      if @signOfInequality == ""
-        output += getMaxAndTotalValueResultStirng(maxValue, totalValue, totalDiceCount)
+    result =
+      if @cmp_op
+        success_count = roll_list.count do |e|
+          x = e[:sum] + @modify_number
+          # Ruby 1.8のケア
+          @cmp_op == :'!=' ? x != @target_number : x.send(@cmp_op, @target_number)
+        end
+        "成功数#{success_count}"
+      else
+        sum_list = roll_list.map { |e| e[:sum] }
+        total = sum_list.inject(0, :+) + @modify_number
+        max = sum_list.map { |i| i + @modify_number }.max
+        "#{max}/#{total}(最大/合計)"
       end
-    end
 
-    return output
-  end
+    sequence = [
+      "#{@nick_e}: (#{expr()})",
+      dice_text(roll_list) + format_modifier(@modify_number),
+      result
+    ]
 
-  def getMaxAndTotalValueResultStirng(maxValue, totalValue, _totalDiceCount)
-    return " ＞ #{maxValue}/#{totalValue}(最大/合計)"
+    return sequence.join(" ＞ ")
   end
 
-  def getAddRollUpperTarget(target1, target2)
-    if  target1
-      return target1.to_i
-    end
-
-    if  target2
-      return target2.to_i
-    end
+  private
 
+  # ダイスロールし、ダイスボットのソート設定に応じてソートする
+  #
+  # @param times [Integer] ダイスの個数
+  # @param sides [Integer] ダイスの面数
+  # @return [Array<Hash>]
+  def roll(times, sides)
     if @diceBot.upplerRollThreshold == "Max"
-      return 2
-    else
-      return @diceBot.upplerRollThreshold
+      @reroll_threshold = sides
     end
-  end
 
-  # 入力の修正値の部分からボーナスの数値に変換する
-  # @param [String] modifier 入力の修正値部分
-  # @return [Integer] ボーナスの数値
-  def getBonusValue(modifier)
-    if modifier.empty?
-      0
-    else
-      ArithmeticEvaluator.new.eval(modifier, @diceBot.fractionType.to_sym)
+    ret = Array.new(times) { roll_ones(sides) }
+    ret.map! { |e| {:sum => e.inject(0, :+), :list => e} }
```

Comment:
Array.new の中だけで作れました。

```suggestion
    ret = Array.new(times) do
      list = roll_ones(sides)
      { :sum => list.inject(0, :+), :list => list }
    end
```